cmd/logger.go

```diff
@@ -137,7 +93,58 @@ func getSource() string {
 	return fmt.Sprintf("[%s:%d:%s()]", filename, lineNum, funcName)
 }
 
-func logIf(level logrus.Level, source string, err error, msg string, data ...interface{}) {
+// TracePoint is to capture stack trace
+type TracePoint struct {
+	Line     int    `json:"line,omitempty"`
+	Filename string `json:"file,omitempty"`
+	Function string `json:"func,omitempty"`
+}
+
+// String method - returns the string
+// representation of a TracePoint instance
+func (tp TracePoint) String() string {
+	return tp.Filename + ":" + strconv.Itoa(tp.Line) + ":" + tp.Function
+}
+
+// trace method - to create stack trace
+func trace(level int) []string {
+	var stack []string
+	var pc uintptr
+	var file string
+	var line int
+	ok := true
+	var callTrace []TracePoint
+	pc, file, line, ok = runtime.Caller(level)
+	rootPath := filepath.Dir(file)
+	for ok {
+		// Clean up the common root path, rootPath,
+		// from the collected stack trace
+		function := runtime.FuncForPC(pc).Name()
+		_, function = filepath.Split(function)
+		file = strings.TrimPrefix(file, rootPath+string(os.PathSeparator))
+		tp := TracePoint{}
+		tp = TracePoint{Line: line, Filename: file, Function: function}
+		callTrace = append(callTrace, tp)
+		level++
+		pc, file, line, ok = runtime.Caller(level)
```

Comment:
For below call trace, what is the expected trace output?

`github.com/minio/minio/cmd/utils.go:1234:myfunc()` calls 
`github.com/minio/minio/pkg/mypkg/myutil.go:12:Myutilfunc()` calls
`github.com/myuser/myproject/mypkg/checker.go:34:Mycheck()`

I suspect `file` is not cleaned up properly.

Reply:
The output would be like:
```
utils.go:1234:myfunc()
github.com/minio/minio/pkg/mypkg/myutil.go:12:Myutilfunc()
github.com/myuser/myproject/mypkg/checker.go:34:Mycheck()
```
where trimmed `rootPath` is set as:
```
rootPath := filepath.Dir(file) + string(os.PathSeparator),
```
and if it comes out to be `github.com/minio/minio/cmd`.


A more realistic example looks like below:
```
Trace: /home/ersan/go/src/net/http/server.go:2619:http.serverHandler.ServeHTTP
       /home/ersan/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP
       /home/ersan/work/src/github.com/minio/minio/pkg/http/server.go:107:http.(*Server).Start.func1
       <autogenerated>:1:cmd.(*reservedMetadataHandler).ServeHTTP
       generic-handlers.go:130:cmd.reservedMetadataHandler.ServeHTTP
       <autogenerated>:1:cmd.(*authHandler).ServeHTTP
       auth-handler.go:234:cmd.authHandler.ServeHTTP
       <autogenerated>:1:cmd.(*resourceHandler).ServeHTTP
       generic-handlers.go:445:cmd.resourceHandler.ServeHTTP
       /home/ersan/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP
       /home/ersan/work/src/github.com/minio/minio/vendor/github.com/rs/cors/cors.go:190:cors.(*Cors).Handler.func1
       <autogenerated>:1:cmd.(*timeValidityHandler).ServeHTTP
       generic-handlers.go:334:cmd.timeValidityHandler.ServeHTTP
       <autogenerated>:1:cmd.(*cacheControlHandler).ServeHTTP
       generic-handlers.go:243:cmd.cacheControlHandler.ServeHTTP
       <autogenerated>:1:cmd.(*minioReservedBucketHandler).ServeHTTP
       generic-handlers.go:264:cmd.minioReservedBucketHandler.ServeHTTP
       <autogenerated>:1:cmd.(*redirectHandler).ServeHTTP
       generic-handlers.go:216:cmd.redirectHandler.ServeHTTP
       <autogenerated>:1:cmd.(*crossDomainPolicy).ServeHTTP
       crossdomain-xml-handler.go:51:cmd.crossDomainPolicy.ServeHTTP
       <autogenerated>:1:cmd.(*requestHeaderSizeLimitHandler).ServeHTTP
       generic-handlers.go:87:cmd.requestHeaderSizeLimitHandler.ServeHTTP
       <autogenerated>:1:cmd.(*requestSizeLimitHandler).ServeHTTP
       generic-handlers.go:62:cmd.requestSizeLimitHandler.ServeHTTP
       <autogenerated>:1:cmd.(*httpStatsHandler).ServeHTTP
       generic-handlers.go:494:cmd.httpStatsHandler.ServeHTTP
       <autogenerated>:1:cmd.(*pathValidityHandler).ServeHTTP
       generic-handlers.go:555:cmd.pathValidityHandler.ServeHTTP
       /home/ersan/work/src/github.com/minio/minio/vendor/github.com/gorilla/mux/mux.go:107:mux.(*Router).ServeHTTP
       /home/ersan/go/src/net/http/server.go:1918:http.HandlerFunc.ServeHTTP
       api-router.go:63:cmd.(objectAPIHandlers).PutObjectHandler-fm
       object-handlers.go:483:cmd.objectAPIHandlers.PutObjectHandler
Source: [object-handlers.go:483:objectAPIHandlers.PutObjectHandler()]
Time: 2017-12-28T00:01:02.096951197Z
Cause: illegal base64 data at input byte 4
ERROR: Unable to validate content-md5 format.
```
where `rootPath = /home/ersan/work/src/github.com/minio/minio/cmd/`

Reply:
> utils.go:1234:myfunc()
> github.com/minio/minio/pkg/mypkg/myutil.go:12:Myutilfunc()
> github.com/myuser/myproject/mypkg/checker.go:34:Mycheck()

Yes. This looks odd to me. `github.com/minio/minio/pkg/mypkg/myutil.go` is not cleaned up.  I think we would need to trim source path like
1. trim `github.com/minio/minio` from all minio source file paths. this will make like below
```
main:go:43:...
cmd/utils.go:1234:...
pkg/mypkg/myutl.go:12:...
...
```
2. I am not sure whether we can trim `github.com` from other sources.  I think we need to preserve.

You can discuss with @abperiasamy and close it accordingly.

Reply:
Take a look at errors.Trace which provides similar functionality.